node/forest_libp2p/src/blocksync/provider.rs

```diff
@@ -0,0 +1,241 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use chain::Error as ChainError;
+use forest_blocks::{Tipset, TipsetKeys};
+use forest_cid::Cid;
+use ipld_blockstore::BlockStore;
+use log::debug;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use super::{
+    BlockSyncRequest, BlockSyncResponse, BlockSyncResponseStatus, CompactedMessages, TipsetBundle,
+};
+
+/// Builds blocksync response out of chain data.
+pub fn make_blocksync_response<DB>(db: Arc<DB>, request: &BlockSyncRequest) -> BlockSyncResponse
+where
+    DB: BlockStore,
+{
+    let mut response_chain: Vec<TipsetBundle> = vec![];
```

Comment:
```suggestion
    let mut response_chain: Vec<TipsetBundle> = Vec::with_capacity(request.request_len as usize);
```
The bundles will be of non-trivial size, best to avoid reallocations as it's rare for partial responses. (If you want, you could also just set capacity to the max of the request length or epoch)

Reply:
updated